File: 18_자율문제/6월4주차/박지원/pgopenchat.js
//프로그래머스 오픈채팅방 자바스크립트
//무난 무난한 문제 근데 아래 다른 사람이 푼 방식이 자스에선 더 좋을듯!

//내가푼방식
function solution(record) {
    var answer = [];
    let userMap = new Map();
    record.map((el) => {
        const records = el.split(" ");
        if (records[0] != 'Leave') {
            userMap.set(records[1], records[2]);
        }
    })
    record.map((el) => {
        const records = el.split(" ");
        if (records[0] == 'Enter') {
            answer.push(`${userMap.get(records[1])}님이 들어왔습니다.`);
        } else if (records[0] == 'Leave') {
            answer.push(`${userMap.get(records[1])}님이 나갔습니다.`)
        }
    })
    return answer;
}

//다른 사람이 푼 방식
function solution(record) {
    const userInfo = {};
    const action = [];
    const stateMapping = {
        'Enter': '님이 들어왔습니다.',
        'Leave': '님이 나갔습니다.'
    }

    record.forEach((v) => {
        const [state, id, nick] = v.split(' ');

        if(state !== "Change") {
            action.push([state, id]);
        }

        if(nick) {
            userInfo[id] = nick;
        }
    })

    return action.map(([state, uid]) => {
        return `${userInfo[uid]}${stateMapping[state]}`;    
    })
}

File: 18_자율문제/2024_상반기/6월4주차/박지원/pgopenchat.java
//프로그래머스 오픈채팅방 자바
//풀이는 똑같은데.. map.put이랑 문자열 비교인 equals 때문에 한~~~참을 헤멨다..

import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        HashMap<String, String> map = new HashMap<>();
        int cnt = 0;
        for (int i = 0; i < record.length; i++) {
            String[] arr = record[i].split(" ");
            if (!arr[0].equals("Leave")) {
                map.put(arr[1], arr[2]);
            }
            if (!arr[0].equals("Change")) cnt++;
        }
        String[] answer = new String[cnt];
        int idx = 0;
        for (int i = 0; i < record.length; i++) {
            String[] arr = record[i].split(" ");
            String str = map.get(arr[1]);
            if (arr[0].equals("Enter")) {
                answer[idx] = str + "님이 들어왔습니다.";
                idx++;
            } else if (arr[0].equals("Leave")) {
                answer[idx] = str + "님이 나갔습니다.";
                idx++;
            }
        
        }
        return answer;
    }
}
